HPCC-20164 Add WsDfuEx.DFUCreateAndPublish

wangkx opened the pull request:
This method is used to create a logical file using ECL Record
Definition. The input for this method includes: logical file
name, group name, JobID, and ECL Record Definition. The outout
is file part information on cluster.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx commented:
@rpastrana please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 128626b998aac008d671cfe08efa3d04a20515b1
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |804 |804 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  173 sec (00:02:53) |   61 sec (00:01:01) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1301 sec (00:21:41) |   18 sec (00:00:18) | 1597 sec (00:26:37) |

rpastrana commented:
@ghalliday we're unclear about a) the need to associate the newly created file with a rec def (it seems necessary) b) how to associate an ecl rec def (currently setting ecl(), but not able to confirm the rec def has actually been associated w/ the file)
Also, the file doesn't report correctly on eclwatch due to invalid/unavailable file size (perhaps expected due to the nature of the file).

ghalliday commented:
@wangkx @rpastrana the code looks sensible to me - I didn't spot any problems.
I think that is the correct way of associating the record.  Really Jake is the expect on this area. I had a few questions.
- What happens if the number of file parts doesn't match the size of the cluster?  Or will we require spark to generate the correct number of parts.
- Should there be scope accessing checking for a write operation?
- @richardkchapman will there be any problem not having the new type info.  I think it will derive it from the record instead.
- Not sure if number of copies is correct - does it need to be set to 1?

richardkchapman commented:
It should create the new record info from the ECL, but might be sensible to do that here rather than leaving all users of the file to do it (would also server as a way to check that the ECL passed in was valid)

rpastrana commented:
@ghalliday
- 1-to-1 file part to thor node mapping is expected, but this service will return file part information so we'll be able to handle n-to-m mapping...  can we dictate number of file parts for the newly created file at this point? 
- Scope authorization seem like a requirement. @wangkx does this code uphold scope auth?

wangkx commented:
ESP never checks file scope. It is done on dali.

richardkchapman commented:
Needs rebasing. And we do need the scope authorization to be done, even if ESP gets Dali to do it...

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 57a3e32b38b663adf346527dce02b7c391c44046
Build: success
Build: failed!
Number of Error(s): 2
CPack Error: Error when generating package: hpccsystems-platform

[Build error:](https://gist.github.com/31321eb29b5082903903bc0768bb0182)
Build log [Build log](https://gist.github.com/HPCCSmoketest/31321eb29b5082903903bc0768bb0182#file-result-18-08-07-10-12-43-log)

Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  170 sec (00:02:50) |    4 sec (00:00:04) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  183 sec (00:03:03) |

wangkx commented:
@rpastrana please review my changes.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 4e141f960914d6eae01ab01c30fc2f4bf4bf9ad2
Build: success
Build: failed!
Number of Error(s): 2
CPack Error: Error when generating package: hpccsystems-platform

[Build error:](https://gist.github.com/96390b1aa80725d55e32890e30aeb032)
Build log [Build log](https://gist.github.com/HPCCSmoketest/96390b1aa80725d55e32890e30aeb032#file-result-18-08-08-17-23-32-log)

Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  174 sec (00:02:54) |    3 sec (00:00:03) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |    0 sec (00:00:00) |  194 sec (00:03:14) |

wangkx commented:
@AttilaVamos build error: HPCC Uninstall: failed (Missing 'complete-uninstall.sh' file.)

rpastrana commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+        for (unsigned int i=0; i<part.numCopies(); i++)
+        {
+            StringBuffer b;
+            part.queryNode(i)->endpoint().getUrlStr(b);
```
Outside of the scope of this issue, but the Spark-hpcc code exposes an issue w/ the url reported for each file part, there's situations where internal ips are reported which requires clients to somehow map to externally accessible addresses. Might it be possible to request all known ips for each file part? @jakesmith ? If not, we might need to provide a separate mechanism for mapping file parts based on internal ips to external ips...

jakesmith replied:
I'm not sure, as you say beyond the scope of this issue. @rpastrana - Is there a separate JIRA open to discuss?
I don't know how typical a setup it is, or how broad an issue it is in general, i.e. apart from effecting this file part url info.
Should there be a general service to request the external version of a named group [somehow] ? I think the environment (xml) would need extending to define exactly what the relationship is too.

rpastrana replied:
@jakesmith https://track.hpccsystems.com/browse/HPCC-20308

rpastrana commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+
+            Owned<IEspDFUPart> filePart = createDFUPart("","");
+            filePart->setId(1 + part.queryPartIndex());
+            filePart->setPartSizeInt64(size);
```
is there a non Int64 partsize we need to set as well?

jakesmith replied:
not new, but including the type in the method name seems a little odd, i.e. why is it not called setPartSize

rpastrana reviewed (commented):
@wangkx a few more comments/questions. we should also have @jakesmith review.

wangkx commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+    file->attach(logicalName, userDesc);
+}
+
+void CWsDfuEx::getFilePartsOnCluster(IEspContext &context, const char *logicalName, const char *groupName,
```
There is no  ::onDFUFilePartsOnCluster method. But, similar logic is in onDFUInfo.

wangkx commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+        const char *recordDefinition = req.getECLRecordDefinition();
+        if (isEmptyString(logicalName))
+             throw MakeStringException(ECLWATCH_INVALID_INPUT,"No LogicalName defined.");
+        if (isEmptyString(jobID))
```
It is not DFUWUID. It is a meaningful name to identify the job.

rpastrana replied:
Ok, in that case, this should be optional. 
Unless it's commonly done differently, this method should set the jobID to something like "CreateAndPusblish <filename>" maybe "CreateAndPublish <group::filename>

wangkx commented on esp/services/ws_dfu/ws_dfuService.hpp:
```diff
@@ -231,6 +232,10 @@ class CWsDfuEx : public CWsDfu
     void queryFieldNames(IEspContext &context, const char *fileName, const char *cluster,
         unsigned __int64 fieldMask, StringArray &fieldNames);
     void parseFieldMask(unsigned __int64 fieldMask, unsigned &fieldCount, IntArray &fieldIndexArray);
+    void createAndPublishLogicalFile(IEspContext &context, const char *logicalName, const char *groupName,
```
It is private.

rpastrana commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+
+    //Create a FileDescriptor
+    Owned<IFileDescriptor> fileDesc = createFileDescriptor();
+    fileDesc->queryProperties().setProp("@job", jobID);
```
see comment above about generated jobid

jakesmith replied:
@rpastrana  I think the intention was that the job id would be something created by the client (e.g. Spark) that was unique to it.

rpastrana replied:
right, so then the jobid (let's call it something different btw) should be optional. In the absence of a client provided entry, we should default to a meaningful job description...

jakesmith replied:
I'm not sure it should go into the file descriptor at all.
What I thought this was was a unique job ID created by the client, akin to workunit, that would be used as part of the token and signed, so it was verifiably unaltered by the time the request got to the dafilesrv's.

rpastrana replied:
perhaps it is (based on the name). I was told this morning, that this attribute is displayed in ECLWatch as "job name" and is a human readable job description...

wangkx replied:
In ECLWatch, it shows inside the 'Job Name' field.

rpastrana replied:
@wangkx,  @jakesmith clarified this issue.
There are essentially 2 jobid fields in this method. One is the dfufile attribute which ends up being the job description. As mentioned before that should be a human readable description, and can default to "CreateAndPublish file <filename>".

The access token should also incorporate a "reaquestid" to increase uniqueness of the token.

jakesmith commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+
+    //Publish the file
+    Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fileDesc);
+    file->setAccessed();
```
I'm surprised attach doesn't effectively do this..

jakesmith commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,6 +812,19 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+    string LogicalName;
+    string GroupName;
+    string JobID;
+    string ECLRecordDefinition;
```
or (as mentioned elsewhere), could we just use the existing DFUInfo return structure which has everything we need?

wangkx replied:
I am going to use the existing DFUInfo return structure.

rpastrana replied:
@wangkx based on today's meeting, the access token will require a file access level as well, for this method we might always set the access level to WRITE or perhaps FULL

rpastrana replied:
we also need an optional "JobDescription" field.

jakesmith commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,182 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+        createAndPublishLogicalFile(context, logicalName, groupName, jobID, userID.str(), recordDefinition, userDesc);
+
+        IEspDFUFilePartsOnCluster &partsOnCluster = resp.updateParts();
+        getFilePartsOnCluster(context, logicalName, groupName, userDesc, partsOnCluster);
```
as mentioned elsewhere, could this new service not be combined with DFUInfo ?
Isn't essentially what is happening here is a DFUInfo reply having 1st created a new file?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 13eaef2a2d27a40480e895f7aa886156e6532e18
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   32 sec (00:00:32) |  174 sec (00:02:54) |   61 sec (00:01:01) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1318 sec (00:21:58) |   18 sec (00:00:18) | 1638 sec (00:27:18) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 04353a555e12c5062c8815cc0d537c76c088e588
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  174 sec (00:02:54) |   61 sec (00:01:01) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1264 sec (00:21:04) |   19 sec (00:00:19) | 1570 sec (00:26:10) |

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,6 +812,23 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+    string ECLRecordDefinition;
+    bool ReturnJsonTypeInfo(false);
+    bool ReturnBinTypeInfo(false);
+    int AccessTokenExpirationSeconds;
```
We need a flag to request the AccessToken as well

rpastrana commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -5885,4 +5885,115 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
+            userDesc->set(userID.str(), context.queryPassword(), context.querySignature());
+        }
+
+        const char *jobID = req.getJobID();
```
there will now be a new input field "jobdescription" to be used here instead of jobid.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: ba2ef93947df09a7ff8aeacf87655a3d5cff48e6
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  178 sec (00:02:58) |   61 sec (00:01:01) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1298 sec (00:21:38) |   18 sec (00:00:18) | 1610 sec (00:26:50) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 54cadf7af80ba2f8e4699e1d580b28d9396224d6
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   38 sec (00:00:38) |  172 sec (00:02:52) |   62 sec (00:01:02) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1314 sec (00:21:54) |   18 sec (00:00:18) | 1639 sec (00:27:19) |

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,6 +821,30 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+{
+    string LogicalName;
+    string Cluster;
+    string JobDescription;
```
let's remove and always set to "CreateAndPublish file 'filename'"

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,6 +821,30 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+    
+    bool ReturnJsonTypeInfo(false);
+    bool ReturnBinTypeInfo(false);
+    ESPenum SecAccessType AccessType;
```
We might not need this unless @jakesmith 's dafileserv will require the user to provide "Write" access as part of the request.

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,6 +821,30 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+    ESPenum SecAccessType AccessType; //Copy from request
+    int AccessTokenExpirationMinutes; //Copy from request if < 24hrs. Max 24hrs
+    ESPstruct DFUFileDetail FileDetail;
+    string AccessToken;
```
@jakesmith i think expressed an issue w/ this name, please provide an accurate alternative.

rpastrana replied:
Also, this security artifact will follow @jakesmith new format

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,6 +821,30 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+    string JobDescription; //Copy from request
+    ESPenum SecAccessType AccessType; //Copy from request
+    int AccessTokenExpirationMinutes; //Copy from request if < 24hrs. Max 24hrs
+    ESPstruct DFUFileDetail FileDetail;
```
this will follow @jakesmith new abbreviated fileparts structure

wangkx committed:
HPCC-20340 Add WsDfu.DFUReadAccess

This new method is used to get signed access to logical files.
It also returns meta data intended to be used by clients to
read data.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx committed:
HPCC-20164 Add WsDfu.DFUCreateAndPublish

This method is used to create/publish a logical file using ECL
Record Definition. The input for this method includes: logical
file name, cluster name, ECL Record Definition, etc. The output
is file information on cluster and assess token.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 0ee83a4fc4295a9d4684e6bf9666f734463a7875
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  178 sec (00:02:58) |   92 sec (00:01:32) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1333 sec (00:22:13) |   17 sec (00:00:17) | 1699 sec (00:28:19) |

wangkx committed:
HPCC-20340 Add 2 changes into DFUReadAccess interface

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: bd51f06ea9ba971756fe9258b5520d8c701dad95
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   37 sec (00:00:37) |  172 sec (00:02:52) |   94 sec (00:01:34) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1327 sec (00:22:07) |   20 sec (00:00:20) | 1712 sec (00:28:32) |

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,7 +821,67 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+    string Name; //Copy from request
+    string UserID; //the uesr of the DFUCreateAndPublishRequest
+    ESPenum SecAccessType AccessType; //Copy from request
+    int ExpiryMinutes; //Copy from request if < 24hrs. Max 24hrs
```
this might need to change to secs

rpastrana commented on esp/scm/ws_dfu.ecm:
```diff
@@ -812,7 +821,67 @@ ESPresponse [exceptions_inline, nil_remove] EraseHistoryResponse
+    bool ReturnJsonTypeInfo(false);
+    bool ReturnBinTypeInfo(false);
+    ESPenum SecAccessType AccessType;
+    int ExpiryMinutes(1);
```
this might need to change to secs (60)

rpastrana reviewed (commented):
@wangkx a couple of comments on the interface.

rpastrana commented:
@wangkx per offline discussion:
Createandpublish will need to be broken out into 2 steps (2 wsdfu methods: dfucreateFile, and dfupublishFile)
dfucreateFile: client asks Esp createFile service for file to be created
                        Service: creates a [unattached] file - and serializes all the info. just as it would for a real [attach] file

client uses that info to write to physical parts (in the end when dafilesrv supports protocol, the info will be like read, client->dafilesrv will says write to part N)

dfupublishFile: client asks Esp publishFile service to publish "blah", where blah contains enough info for the Esp service to recreate the temp logical file, rename parts and attach.
The "blah" might just be the name of the logical file and the service just maps that to temp name/parts in a consistent way.

wangkx commented:
A new PR will be created to replace this PR.
